fs/fs_zenfs.cc

```diff
@@ -1304,7 +1317,122 @@ void ZenFS::GetZenFSSnapshot(ZenFSSnapshot& snapshot,
+    }
+  }
+
+  for (const auto& it : file_extents) {
+    s = MigrateFileExtents(it.first, it.second);
+    if (!s.ok()) break;
+  }
+  return s;
+}
+
+IOStatus ZenFS::MigrateFileExtents(
+    const std::string& fname,
+    const std::vector<ZoneExtentSnapshot*>& migrate_exts) {
+  IOStatus s = IOStatus::OK();
+  Info(logger_, "MigrateFileExtents, fname: %s, extent count: %lu",
+       fname.data(), migrate_exts.size());
+  // The file may be deleted by other threads, better double check.
+  auto zfile = GetFile(fname);
+  if (zfile == nullptr || zfile->IsOpenForWR()) {
+    return IOStatus::OK();
+  }
+
+  std::vector<ZoneExtent*> new_extent_list;
+  std::vector<ZoneExtent*> extents = zfile->GetExtents();
+  for (const auto* ext : extents) {
+    new_extent_list.push_back(
+        new ZoneExtent(ext->start_, ext->length_, ext->zone_));
+  }
+
+  // Modify the new extent list
+  for (ZoneExtent* ext : new_extent_list) {
+    // Check if current extent need to be migrated
+    auto it = std::find_if(migrate_exts.begin(), migrate_exts.end(),
+                           [&](const ZoneExtentSnapshot* ext_snapshot) {
+                             return ext_snapshot->start == ext->start_ &&
+                                    ext_snapshot->length == ext->length_;
+                           });
+
+    if (it == migrate_exts.end()) {
+      Info(logger_, "Migrate extent not found, ext_start: %lu", ext->start_);
+      continue;
+    }
+
+    Zone* target_zone = nullptr;
+
+    // Allocate a new migration zone.
+    s = zbd_->TakeMigrateZone(&target_zone, zfile->GetWriteLifeTimeHint(),
+                              ext->length_);
+    if (!s.ok()) {
+      continue;
+    }
+
+    if (target_zone == nullptr) {
+      zbd_->ReleaseMigrateZone(target_zone);
+      Info(logger_, "Migrate Zone Acquire Failed, Ignore Task.");
+      continue;
+    }
+
+    uint64_t target_start = target_zone->wp_;
+    if (zfile->IsSparse()) {
+      // For buffered write, ZenFS use inlined metadata for extents and each
+      // extent has a SPARSE_HEADER_SIZE.
+      target_start = target_zone->wp_ + ZoneFile::SPARSE_HEADER_SIZE;
+      zfile->MigrateData(ext->start_ - ZoneFile::SPARSE_HEADER_SIZE,
+                         ext->length_ + ZoneFile::SPARSE_HEADER_SIZE,
+                         target_zone);
+    } else {
+      zfile->MigrateData(ext->start_, ext->length_, target_zone);
+    }
+
+    // If the file doesn't exist, skip
+    if (GetFileInternal(fname) == nullptr) {
+      Info(logger_, "Migrate file not exist anymore.");
+      zbd_->ReleaseMigrateZone(target_zone);
+      break;
+    }
+
+    ext->start_ = target_start;
+    ext->zone_ = target_zone;
+    ext->zone_->used_capacity_ += ext->length_;
+
+    zbd_->ReleaseMigrateZone(target_zone);
+  }
+
+  // Check again if the file still exist
+  if (GetFileInternal(fname) == nullptr) {
+    Info(logger_, "Migrate file not exist anymore.");
+    return IOStatus::OK();
+  }
+
```

Comment:
The check does not guarantee that we won't get a file delete before the metadata record is written,  thereby risking creating an invalid record for a file that was previously deleted. We need to do the check holding the file list mutex, preventing any deletes while we are updating the extents in the meta log.

Reply:
Understand, I thought an invalid record log doesn't hurt anyone (we may barely meet this in production), so I tried to avoid a lock here on purpose, LOL, do you think we should lock here anyway?

Reply:
Fixed anyway